### PR TITLE
Closes #3245: Add `poisson` distribution to random number generators

### LINKED ISSUE #3245: add `poisson` to random number generators
add poisson to random number generators 

Relevant docs:
[numpy.random.Generator.poisson](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.poisson.html)

### PR DESCRIPTION
**Pull request by stress-tess**
This PR (closes #3245) adds the `poisson` distribution to random number generators. This also adds the docs and testing including hypothesis testing against the probabilities expected from the poisson probability mass function. This PR does a coforall over the locals and tasks so I can ensure each task gets a unique seed that is reproducable

I also moved one of the other hypothesis tests so they are all together

### CODE REVIEW

File: src/RandMsg.chpl [hunk 1]
```diff
@@ -10,6 +10,7 @@ module RandMsg
     use Logging;
     use Message;
     use RandArray;
+    use Unique;
```
Comment (jeremiah-corrado) [inline src/RandMsg.chpl:13]: I'm guessing this is no longer needed?

File: src/RandMsg.chpl [hunk 2]
```diff
@@ -547,6 +548,101 @@ module RandMsg
...
+        // https://en.wikipedia.org/wiki/Poisson_distribution#Random_variate_generation
+        // generates values drawn from poisson distribution using a stream of uniformly distributed random numbers
+
+        const nTasksPerLoc = here.maxTaskPar; // tasks per locale based on locale0
+        const Tasks = {0..#nTasksPerLoc}; // these need to be const for comms/performance reasons
+
+        const generatorSeed = (rng.next() * 2**62):int;
+        var poissonArr = makeDistArray(size, int);
+
+        // I hate the code duplication here but it's not immediately obvious to me how to avoid it
```
Comment (jeremiah-corrado) [inline src/RandMsg.chpl:583]: Not sure if it's worth all the overhead, but you could use a record type that can store either a single lambda value, or an array of lambdas s.t. `const lam = lamArr[locSubDom.low + i];` always works:
```chapel
record lambda {
  var isArray: bool;
  var arr;
  var val: real;

  proc init(arr: [] real) {
    this.isArray = true;
    this.arr = arr;
  }

  proc init(val) {
    this.isArray = false;
    this.arr = makeDistArray(0..0);
    this.val = val:real;
  }

  proc this(idx) {
    return if isArray then this.arr[idx] else this.val;
  }
}
```

And then initialize it differently depending on the value of `isSingleLam`:
```chapel
var lamArr = if isSingleLam then new lambda(lamStr)
  else new lambda(toSymEntry(getGenericTypedArrayEntry(lamStr, st),real).a);
```

**Review by jeremiah-corrado — APPROVED** [review]
Looks good! Just a couple of suggestions:

File: src/RandMsg.chpl (hunk 1 above)
Comment (stress-tess) [reply src/RandMsg.chpl:13]: oops! yep ur right!

File: src/RandMsg.chpl (hunk 2 above)
Comment (stress-tess) [reply src/RandMsg.chpl:582]: oh yah I like this a lot! thanks for this! I bet theres quite a few places in the codebase that benefit from something like this

**Review by jaketrookman — APPROVED** [review]

File: PROTO_tests/tests/random_test.py [hunk 3]
```diff
@@ -246,6 +264,33 @@ def test_normal_hypothesis_testing(self):
...
+        while (1 - sp_stats.poisson.cdf(num_elems, mu=lam)) > tol:
+            num_elems += 5
+
+        obs_counts = np.array([0] * num_elems)
+        for k, v in count_dict.items():
+            obs_counts[k] = v
+
+        # use the probability mass function to get the probability of seeing each value
+        # and multiply by num_samples to get the expected counts
+        exp_counts = np.array([sp_stats.poisson.pmf(v, mu=lam) for v in range(num_elems)]) * num_samples
```
Comment (ajpotts) [inline PROTO_tests/tests/random_test.py:290]: This could also be written `sp_stats.poisson.pmf(range(num_elems), mu=lam) * num_samples`.

**Review by ajpotts — APPROVED** [review]
Looks good.  Thanks!

**Pull request merged**